plugins/binanceExchange_ws.go

```diff
@@ -0,0 +1,283 @@
+}
+
+type errConversion struct {
+	from string
+	to   string
+}
+
+func (err errConversion) Error() string {
+	return fmt.Sprintf("Error conversion from %s to %s", err.from, err.to)
+}
+
+type stream struct {
+	doneC   chan struct{}
+	stopC   chan struct{}
+	cleanup func()
+}
+
+//Wait until the stream ends
+func (s stream) Wait() {
+
+	if s.doneC == nil {
+		return
+	}
+
+	<-s.doneC
+}
+
+//Close the stream and cleanup any data
+func (s stream) Close() {
+	if s.stopC == nil {
+		return
+	}
+	s.stopC <- struct{}{}
+	s.stopC = nil
+
+	if s.cleanup != nil {
+		s.cleanup()
+	}
+}
+
+type mapData struct {
+	data      interface{}
+	createdAt time.Time
+}
+
+func isStale(data mapData, ttl time.Duration) bool {
+
+	return time.Now().Sub(data.createdAt).Seconds() > ttl.Seconds()
+}
+
+type mapEvents struct {
+	data map[string]mapData
+	mtx  *sync.Mutex
+}
+
+func (m *mapEvents) Set(key string, data interface{}) {
+
+	now := time.Now()
+
+	m.mtx.Lock()
+	m.data[key] = mapData{
+		data:      data,
+		createdAt: now,
+	}
+	m.mtx.Unlock()
+}
+
+func (m *mapEvents) Get(key string) (mapData, bool) {
+	m.mtx.Lock()
+	data, isData := m.data[key]
+	m.mtx.Unlock()
+
+	return data, isData
+}
+
+func (m *mapEvents) Del(key string) {
+	m.mtx.Lock()
+	delete(m.data, key)
+	m.mtx.Unlock()
+
+}
+
+func makeMapEvents() *mapEvents {
+	return &mapEvents{
+		data: make(map[string]mapData),
+		mtx:  &sync.Mutex{},
+	}
+}
+
+type events struct {
+	SymbolStats *mapEvents
+}
+
+func createStateEvents() *events {
+	events := &events{
+		SymbolStats: makeMapEvents(),
+	}
+
+	return events
+}
+
+// subscribe for symbol@ticker
+func subcribeTicker(symbol string, state *mapEvents) (*stream, error) {
+
+	wsMarketStatHandler := func(ticker *binance.WsMarketStatEvent) {
+		state.Set(symbol, ticker)
+	}
+
+	errHandler := func(err error) {
+		log.Printf("Error WsMarketsStat for symbol %s: %v\n", symbol, err)
+	}
+
+	doneC, stopC, err := binance.WsMarketStatServe(symbol, wsMarketStatHandler, errHandler)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &stream{doneC: doneC, stopC: stopC, cleanup: func() {
+		state.Del(symbol)
+	}}, err
+
+}
+
+type binanceExchangeWs struct {
+	events    *events
+	delimiter string
+
+	streams map[string]*stream
+
+	assetConverter model.AssetConverterInterface
+
+	mtx *sync.Mutex
+}
+
+// makeBinanceWs is a factory method to make an binance exchange over ws
+func makeBinanceWs() (*binanceExchangeWs, error) {
+
+	binance.WebsocketKeepalive = true
+
+	events := createStateEvents()
+
+	beWs := &binanceExchangeWs{
+		events:         events,
+		delimiter:      "",
+		assetConverter: model.CcxtAssetConverter,
+		mtx:            &sync.Mutex{},
+		streams:        make(map[string]*stream),
+	}
+
+	return beWs, nil
+}
+
+func getPrecision(floatStr string) int8 {
+
+	strs := strings.Split(floatStr, ".")
+
+	if len(strs) != 2 {
+		log.Printf("error get precision for float %s\n", floatStr)
```

Comment:
if we are saying it is an error here we should panic. but it's ok to return 0 here and not an error.
can we change the message from "error get precision..." to "could not get precision..."?

this way when we see this log line we are just being informed of this rather than having consumers blaming this log line for some other actual error in the bot.